### PR TITLE
Multi speed fan

### PR DESCRIPTION
**Pull request by wanhanlong1130**
Added multi speed fan using curves

### CODE REVIEW

**Comment by wanhanlong1130** [comment]
Not yet added a test for the new curves. Please don't merge until we check-in on Fri.

File: copper/library.py [hunk 1]
```diff
@@ -43,6 +43,8 @@ def __init__(self, path=chiller_lib, rating_std="", export=False):
                         and not "degradation_coefficient" in p
                         and not "indoor_fan_speeds_mapping" in p
                         and not "indoor_fan_speeds" in p
+                        and not "infdoor_fan_curve_coef" in p
```
Comment (lymereJ) [inline copper/library.py]: I think that there's a typo here, shouldn't it be `indoor_fan_curve_coeff`??

File: tests/test_unitarydirectexpansion.py [hunk 2]
```diff
@@ -258,3 +258,10 @@ def test_get_ranges(self):
         ranges = self.dx_unit_dft.get_ranges()
         assert isinstance(ranges, dict)
         assert len(ranges) == 5
+
+
+# Run the tests
+import unittest
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (lymereJ) [inline tests/test_unitarydirectexpansion.py]: We don't need this block to run the tests. If you have `poetry` installed, you can run `poetry run pytest tests/test_unitarydirectexpansion.py` to run these tests.

File: copper/unitarydirectexpansion.py [hunk 3]
```diff
@@ -202,36 +211,59 @@ def __init__(
...
-                    if (
-                        ratio < capacity_ratio
-                        and capacity_ratios[i + 1] > capacity_ratio
-                    ):
-                        a = (fan_power_fractions[i + 1] - fan_power_fractions[i]) / (
-                            capacity_ratios[i + 1] - capacity_ratios[i]
-                        )
-                        b = fan_power_fractions[i] - a * capacity_ratios[i]
-                        return self.indoor_fan_power * (a * capacity_ratio + b)
+            if self.indoor_fan_curve == 0:
```
Comment (lymereJ) [inline copper/unitarydirectexpansion.py]: Maybe let's initialize `indoor_fan_curve` to `None`.

File: copper/unitarydirectexpansion.py [hunk 4]
```diff
@@ -202,36 +211,59 @@ def __init__(
...
+                            ratio < capacity_ratio
+                            and capacity_ratios[i + 1] > capacity_ratio
+                        ):
+                            a = (
+                                fan_power_fractions[i + 1] - fan_power_fractions[i]
+                            ) / (capacity_ratios[i + 1] - capacity_ratios[i])
+                            b = fan_power_fractions[i] - a * capacity_ratios[i]
+                            return self.indoor_fan_power * (a * capacity_ratio + b)
+            else:  # using curve
+                defualt_coef = 1  # can update this coef later
```
Comment (lymereJ) [inline copper/unitarydirectexpansion.py]: Typo, I assume that it should be `default_coeff`.

File: copper/unitarydirectexpansion.py [hunk 5]
```diff
@@ -202,36 +211,59 @@ def __init__(
...
+                default_min_fan_power = (
+                    self.indoor_fan_power * 0.25
+                )  # default min fan power
+                power_factor = self.default_fan_curve.evaluate(
+                    x=defualt_coef * ratio, y=0
```
Comment (lymereJ) [inline copper/unitarydirectexpansion.py]: Shouldn't x=flow_fraction?

File: copper/unitarydirectexpansion.py [hunk 6]
```diff
@@ -202,36 +211,59 @@ def __init__(
...
+            eqp=self, c_type=self.infdoor_fan_curve_coef["type"]
+        )
+        self.default_fan_curve.coeff1 = self.infdoor_fan_curve_coef["1"]
+        self.default_fan_curve.coeff2 = self.infdoor_fan_curve_coef["2"]
+        self.default_fan_curve.coeff3 = self.infdoor_fan_curve_coef["3"]
+        self.default_fan_curve.coeff4 = self.infdoor_fan_curve_coef["4"]
+
     def calc_fan_power(self, capacity_ratio):
         # Full flow/power
+        ratio = capacity_ratio
```
Comment (lymereJ) [inline copper/unitarydirectexpansion.py]: Is this actually needed?

**Comment by wanhanlong1130** [comment]
@lymereJ  not sure why still have conflicting files; I already tried to merge back from the develop branch and solve conflicts. But, I have fixed the problems you brought up last time and created a new test case for multi speed fan with the default curve.

File: copper/unitarydirectexpansion.py [hunk 7]
```diff
@@ -43,15 +43,23 @@ def __init__(
             "1": {
                 "fan_flow_fraction": 0.66,
                 "fan_power_fraction": 0.4,
-                "capacity_ratio": 0.5,
+                "capacity_fraction": 0.5,
```
Comment (lymereJ) [inline copper/unitarydirectexpansion.py:46]: Why the change from `ratio` to `fraction`? I'm totaly fine with it since they are synonymous but `calc_fan_power` still uses `ratio`. We should probably make the change throughout so it's consistent everywhere.

File: tests/test_unitarydirectexpansion.py [hunk 8]
```diff
@@ -180,6 +180,27 @@ def test_multi_speed(self):
             == 0.65
         )
 
+    def test_multi_speed_with_curve(self):
```
Comment (lymereJ) [inline tests/test_unitarydirectexpansion.py:183]: That's a detail, but if we use a curve we can assume that it's variable-speed (continuous) instead of multi-speed (discrete).

Comment (lymereJ) [reply tests/test_unitarydirectexpansion.py:183]: This test doesn't use a curve, right? So perhaps this is correct and we just need another example for variable-speed that uses the curve.

File: copper/unitarydirectexpansion.py [hunk 9]
```diff
@@ -234,34 +256,47 @@ def calc_fan_power(self, capacity_ratio):
...
-                    if (
-                        ratio < capacity_ratio
-                        and capacity_ratios[i + 1] > capacity_ratio
-                    ):
-                        a = (fan_power_fractions[i + 1] - fan_power_fractions[i]) / (
-                            capacity_ratios[i + 1] - capacity_ratios[i]
-                        )
-                        b = fan_power_fractions[i] - a * capacity_ratios[i]
-                        return self.indoor_fan_power * (a * capacity_ratio + b)
+            if self.indoor_fan_curve == None:
```
Comment (lymereJ) [inline copper/unitarydirectexpansion.py]: If we assign a default curve, this will never be `None`. We perhaps need another way to flag that we need a variable speed fan instead of a multi-speed fan. Maybe something like `indoor_fan_type` where it could either be `constant_volume` or `variable_volume`. If the value is the former then we use the `indoor_fan_speeds` and `indoor_fan_speeds_mapping` and it is the latter we use the new `indoor_fan_curve `.

**Comment by lymereJ** [comment]
I don't think the test failures are due to something you've done. I'll push some changes to try to resolve it.

**Comment by wanhanlong1130** [comment]
@lymereJ  I fixed the two minor points you pointed out last week. If you feel it's correct, can you click "merge pull request"?

**Review by lymereJ — APPROVED** [review]
Looks good, thanks!
